### PR TITLE
Enable CppInterOp

### PR DESCRIPTION
**Pull request by vgvassilev**
[CppInterOp](https://github.com/compiler-research/CppInterOp) exposes API from Clang and LLVM in a mostly backward compatibe way. The API support downstream tools that utilize interactive C++ by using the compiler as a service.

Adopting more of CppInterOp in ROOT will help simplify the LLVM migration process and allow us to upstream more patches to either CppInterOp or LLVM.

@aaronj0, you can take it from here.

### CODE REVIEW

**Comment by vgvassilev** [comment]
Note that if we want to download at configuration time we should use FetchContent_Declare which seems to take similar arguments.

File: core/metacling/src/CMakeLists.txt [hunk 1]
```diff
@@ -74,11 +74,112 @@ if(MSVC)
...
+if(APPLE)
+  set(_cppinterop_extra_cmake_args -DCMAKE_OSX_SYSROOT=${CMAKE_OSX_SYSROOT})
 endif()
 
-add_dependencies(MetaCling CLING)
+if (CMAKE_CXX_STANDARD)
+  list(APPEND _cppinterop_extra_cmake_args -DCMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD})
+endif(CMAKE_CXX_STANDARD)
+
+if (LLVM_FORCE_USE_OLD_TOOLCHAIN)
```
Comment (hahnjo) [inline core/metacling/src/CMakeLists.txt:112]: This is defined in `interpreter/CMakeLists.txt`, so very likely we cannot access it here...

Comment (aaronj0) [reply core/metacling/src/CMakeLists.txt]: We shouldn't need this as we build InterOp with clang>6 and hence gcc>=7.1

**Comment by vgvassilev** [comment]
Note that we need to land https://github.com/compiler-research/CppInterOp/pull/318, release and then merge this PR.

**Comment by vgvassilev** [comment]
Closing in favor or #16814.

**Pull request closed without merging**